src/main/cpp/src/substring_index.cu

```diff
@@ -0,0 +1,208 @@
+        bool const left_to_right = (delimiter_count > 0);
+
+        size_type start_pos = start_char_pos[idx];
+        size_type end_pos   = col_val_len;
+        size_type char_pos  = -1;
+
+        end_char_pos[idx] = col_val_len;
+
+        for (auto i = 0; i < nsearches; ++i) {
+          char_pos = left_to_right ? col_val.find(delim_val, start_pos)
+                                   : col_val.rfind(delim_val, 0, end_pos);
+          if (char_pos == string_view::npos) return;
+          if (left_to_right)
+            start_pos = char_pos + delimiter_len;
+          else
+            end_pos = char_pos;
+        }
+        if (left_to_right)
+          end_char_pos[idx] = char_pos;
+        else
+          start_char_pos[idx] = end_pos + delimiter_len;
+      }
+    });
+}
+
+}  // namespace
+
+template <typename DelimiterItrT>
+std::unique_ptr<column> substring_index(strings_column_view const& strings,
+                                        DelimiterItrT const delimiter_itr,
+                                        size_type count,
+                                        rmm::cuda_stream_view stream,
+                                        rmm::device_async_resource_ref mr)
+{
+  auto strings_count = strings.size();
+  // If there aren't any rows, return an empty strings column
+  if (strings_count == 0) { return make_empty_column(type_id::STRING); }
+
+  // Compute the substring indices first
+  auto start_chars_pos_vec = make_column_from_scalar(numeric_scalar<size_type>(0, true, stream),
+                                                     strings_count,
+                                                     stream,
+                                                     rmm::mr::get_current_device_resource());
+  auto stop_chars_pos_vec  = make_column_from_scalar(numeric_scalar<size_type>(0, true, stream),
+                                                    strings_count,
+                                                    stream,
+                                                    rmm::mr::get_current_device_resource());
+
+  auto start_char_pos = start_chars_pos_vec->mutable_view().data<size_type>();
+  auto end_char_pos   = stop_chars_pos_vec->mutable_view().data<size_type>();
+
+  auto strings_column = column_device_view::create(strings.parent(), stream);
+  auto d_column       = *strings_column;
+
+  // If delimiter count is 0, the output column will contain empty strings
+  if (count != 0) {
+    // Compute the substring indices first
+    compute_substring_indices(
+      d_column, delimiter_itr, count, start_char_pos, end_char_pos, stream, mr);
+  }
+
+  // Extract the substrings using the indices next
+  auto starts_iter =
+    cudf::detail::indexalator_factory::make_input_iterator(start_chars_pos_vec->view());
+  auto stops_iter =
+    cudf::detail::indexalator_factory::make_input_iterator(stop_chars_pos_vec->view());
+  return compute_substrings_from_fn(d_column, starts_iter, stops_iter, stream, mr);
```

Comment:
The original cuDF implementation ([here](https://github.com/rapidsai/cudf/pull/13628)) used `compute_substrings_from_fn` which is in an unnamed namespace, so we need to port that code (along with `substring_from_fn`) to jni if we want to use it. But I'm concerning about maintainability, there is duplicated code, if cuDF changed this code to fix a bug or improve performance, we will forget to update code here too. Could we try to use `cudf::strings::detail::slice_strings` instead of `compute_substrings_from_fn` here?

Reply:
I changed my code in another way.